File: expressjs/src/instrument/middleware.ts
import type { Request, Response, NextFunction } from "express";
import { randomUUID } from "node:crypto";
import type { InstrumentationConfig } from "../config/index.js";
import {
  InstrumenterImpl,
  runWithTraceContext,
  type TraceContext,
} from "./instrument.js";

export function instrumentationMiddleware(config: InstrumentationConfig) {
  return (req: Request, res: Response, next: NextFunction) => {
    // Skip if disabled
    if (!config.enabled) return next();

    // Extract buffer from multi-app context
    const buffer = req.appCtx?.eventBuffer;
    if (!buffer) return next();

    // Sampling
    if (config.sampling_rate < 1.0 && Math.random() > config.sampling_rate) {
      return next();
    }

    const traceId =
      (req.headers["x-trace-id"] as string) || randomUUID();
    const userId = req.user?.id ?? null;

    const instrumenter = new InstrumenterImpl(buffer);
    const ctx: TraceContext = {
      traceId,
      parentSpanId: null,
      userId,
      buffer,
      instrumenter,
    };

    // Set trace ID response header
    res.setHeader("X-Trace-ID", traceId);

    runWithTraceContext(ctx, () => {
      // Create root HTTP span
      const span = instrumenter.startSpan("http", "handler", "request.start");
      span.setMetadata("method", req.method);
      span.setMetadata(
        "path",
        req.originalUrl?.split("?")[0] || req.path,
      );

      // On response finish, end the span
      const onFinish = () => {
        span.setMetadata("status_code", res.statusCode);
        span.setStatus(res.statusCode >= 400 ? "error" : "ok");
        span.end();
        res.removeListener("finish", onFinish);
      };
      res.on("finish", onFinish);

      next();
    });
  };
}

File: golang/internal/instrument/middleware.go
package instrument

import (
	"math/rand"

	"github.com/gofiber/fiber/v2"

	"rocket-backend/internal/config"
	"rocket-backend/internal/metadata"
)

// Middleware returns a Fiber middleware that sets up tracing for each request.
// It generates (or propagates) a trace ID, creates a root HTTP span, and injects
// the instrumenter into the request context for downstream handlers.
// The getBuffer callback extracts the EventBuffer from the request (e.g. from AppContext),
// avoiding a circular dependency between instrument and multiapp packages.
func Middleware(cfg config.InstrumentationConfig, getBuffer func(c *fiber.Ctx) *EventBuffer) fiber.Handler {
	return func(c *fiber.Ctx) error {
		if !cfg.Enabled {
			return c.Next()
		}

		buffer := getBuffer(c)
		if buffer == nil {
			return c.Next()
		}

		// Sampling: skip tracing for a proportion of requests
		if cfg.SamplingRate < 1.0 && rand.Float64() > cfg.SamplingRate {
			return c.Next()
		}

		// Get or generate trace ID from incoming header
		traceID := c.Get("X-Trace-ID")
		if traceID == "" {
			traceID = newUUID()
		}

		// Set up context with trace ID and instrumenter
		ctx := c.UserContext()
		instrumenter := NewInstrumenter(buffer)
		ctx = WithTraceID(ctx, traceID)
		ctx = WithInstrumenter(ctx, instrumenter)
		c.SetUserContext(ctx)

		// Create root HTTP span
		ctx, span := instrumenter.StartSpan(ctx, "http", "handler", "request")
		span.SetMetadata("method", c.Method())
		span.SetMetadata("path", c.Path())
		c.SetUserContext(ctx)

		// Set trace ID response header
		c.Set("X-Trace-ID", traceID)

		// Execute downstream handlers (auth middleware, route handlers, etc.)
		err := c.Next()

		// After downstream completes, extract user ID from auth middleware
		// and attach to the root span (auth middleware sets c.Locals("user"))
		if user, ok := c.Locals("user").(*metadata.UserContext); ok && user != nil {
			span.SetMetadata("user_id", user.ID)
			// Also update the context so any deferred operations have the user ID
			ctx = WithUserID(c.UserContext(), user.ID)
			c.SetUserContext(ctx)
		}

		// Finalize root span with response status
		statusCode := c.Response().StatusCode()
		span.SetMetadata("status_code", statusCode)
		if statusCode >= 400 {
			span.SetStatus("error")
		} else {
			span.SetStatus("ok")
		}
		span.End()

		return err
	}
}
